compiler/rustc_borrowck/src/type_check/mod.rs

```diff
@@ -1057,15 +1057,19 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
     fn check_user_type_annotations(&mut self) {
         debug!(?self.user_type_annotations);
         for user_annotation in self.user_type_annotations {
-            let CanonicalUserTypeAnnotation { span, ref user_ty, inferred_ty } = *user_annotation;
+            let CanonicalUserTypeAnnotation { span, ref user_ty, inferred_ty, variance } =
+                *user_annotation;
             let inferred_ty = self.normalize(inferred_ty, Locations::All(span));
             let annotation = self.instantiate_canonical_with_fresh_inference_vars(span, user_ty);
             match annotation {
                 UserType::Ty(mut ty) => {
                     ty = self.normalize(ty, Locations::All(span));
 
-                    if let Err(terr) = self.eq_types(
+                    // `ty` and `inferred_ty` are swapped for
+                    // better diagnostics.
+                    if let Err(terr) = self.relate_types(
```

Comment:
considering that `ty` and `inferred_ty` now differ, do we have to check for well formedness of `ty` as well?